milestone-2.md

```diff
@@ -0,0 +1,380 @@
+    Identifier(String)
+
+### enum Type
+
+    Base(ScalarType)
+    Array(Box<Type>)
+
+### struct CompilationUnit
+
+    classes: Vec<Class>
+    main: Vec<Stmt>
+
+### struct Class
+
+    id: String
+    members: Vec<ClassMember>
+
+### enum ClassMember
+
+    DataMember {
+        is_static: bool
+        is_private: bool
+        stmt: Stmt
+    }
+
+    MethodMember {
+        is_static: bool
+        is_private: bool
+        return_type: Type
+        id: String
+        params: Vec<Parameter>
+        body: Vec<Stmt>
+    }
+
+    Constructor {
+        id: String
+        params: Vec<Parameter>
+        body: Vec<Stmt>
+    }
+
+### struct Parameter
+
+    type: Type
+    id: String
+
+### enum Stmt
+
+    If {
+        condition: Expr
+        then: Box<Stmt>
+    }
+
+    IfElse {
+        condition: Expr
+        then: Box<Stmt>
+        else: Box<Stmt>
+    }
+
+    While {
+        condition: Expr
+        stmt: Box<Stmt>
+    }
+
+    For {
+        init: Option<Expr>
+        condition: Expr
+        post: Option<Expr>
+        stmt: Box<Stmt>
+    }
+
+    Return(Option<Expr>)
+
+    Cout(Expr)
+
+    Cin(Expr)
+
+    Switch {
+        match_: Expr
+        cases: Vec<Stmt>
+        default: Vec<Stmt>
+    }
+
+    Break
+
+    Expr(Expr)
+
+    Case {
+        case: Expr
+        stmts: Vec<Stmt>
+    }
+
+    Block(Vec<Stmt>)
+
+    VarDeclaration {
+        type_: Type
+        id: String
+        init: Option<Expr>
+    }
+
+### enum Expr
+
+    Null
+
+    This
+
+    Bool(bool)
+
+    Int(i64)
+
+    Char(char)
+
+    Str(String)
+
+    Variable(String)
+
+    Unary {
+        op: Operator
+        rhs: Box<Expr>
+    }
+
+    Binary {
+        lhs: Box<Expr>
+        op: Operator
+        rhs: Box<Expr>
+    }
+
+    Index {
+        id: Box<Expr>
+        index: Box<Expr>
+    }
+
+    ClassAccessor {
+        class: Box<Expr>
+        id: String
+    }
+
+    Function {
+        id: Box<Expr>
+        params: Vec<Expr>
+    }
+
+    NewObject {
+        type_: Type
+        params: Vec<Expr>
+    }
+
+    NewArray {
+        type_: Type
+        length: Box<Expr>
+    }
+
+# Grammar
+
+    Grammar is split into 4 sections: MISC, Expressions, Statements, and CompilationUnit.
+
+    Each section is split into their corresponding rules, and each row represents an option in that rule.
+
+    Rules that contain rules are surrounded by <>. ex: <Expression>
+
+    Terminals are all other groups of words, except a few exceptions that I'm sure you'll understand.
+
+## MISC
+
+#### ScalarType
+
+    void
+    int
+    char
+    string
+    identifier
+
+#### Type #[inline]
+
+    <ScalarType> []*
+
+#### Comma<T> (Takes generic rule T)
+
+    (<T> ,)* <T>?
+
+#### ExpressionList
+
+    <Comma<Expression>>
+
+#### ParameterList
+
+    <Comma<Param>>
+
+## EXPRESSIONS
+
+#### Assignment
+
+    <Or> == <Assignment>
+    <Or> += <Assignment>
+    <Or> -= <Assignment>
+    <Or> \*= <Assignment>
+    <Or> /= <Assignment>
+    Or
```

Comment:
There shouldn't be a need for you to encode precedence in the grammar; your parser combinator should be able to handle it for you.